File: kotlin/infers/src/commonMain/kotlin/dev/stashy/infers/Device.kt
package dev.stashy.infers

/**
 * Live execution device handle.
 */
@SubclassOptInRequired(InfersInternalApi::class)
public interface Device<Self : Device<Self>> {
    public val info: DeviceInfo

    /** Place [hardware] on this device without committing transfer/compute. */
    @InfersInternalApi
    public fun deferHardwareInternal(hardware: HardwareImage): Deferred<Self>
}

public enum class DeviceKind {
    Cpu,
    Gpu,
    Npu,
}

/** Device identity metadata (kind, index, name). */
public data class DeviceInfo(
    public val kind: DeviceKind,
    public val id: ULong,
    public val name: String,
)

/** Singleton CPU execution device. */
public object CpuDevice : Device<CpuDevice> {
    override val info: DeviceInfo = DeviceInfo(DeviceKind.Cpu, 0u, "CPU")

    @OptIn(InfersInternalApi::class)
    override fun deferHardwareInternal(hardware: HardwareImage): Deferred<CpuDevice> {
        hardware.ensureOpen()
        return CpuDeferred(hardware.handle.onCpu())
    }
}

File: crates/infers-bindings/src/device.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, uniffi::Enum)]
pub enum DeviceKind {
    Cpu,
    Gpu,
    Npu,
}

uniffi_mirror! {
    DeviceKind <=> infers_core::DeviceKind,
    [Cpu, Gpu, Npu]
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, uniffi::Record)]
pub struct DeviceInfo {
    pub kind: DeviceKind,
    pub id: u64,
    pub name: String,
}

impl From<infers_core::DeviceInfo> for DeviceInfo {
    fn from(info: infers_core::DeviceInfo) -> Self {
        Self {
            kind: info.kind.into(),
            id: info.id as u64,
            name: info.name,
        }
    }
}

impl From<DeviceInfo> for infers_core::DeviceInfo {
    fn from(info: DeviceInfo) -> Self {
        Self {
            kind: info.kind.into(),
            id: info.id as usize,
            name: info.name,
        }
    }
}

/// CPU execution device (singleton metadata).
#[derive(uniffi::Object)]
pub struct CpuDevice;

#[uniffi::export]
impl CpuDevice {
    #[uniffi::constructor]
    pub fn new() -> std::sync::Arc<Self> {
        std::sync::Arc::new(Self)
    }

    pub fn info(&self) -> DeviceInfo {
        infers_core::Cpu::info().clone().into()
    }
}
